raiden-ts/src/utils/types.ts

```diff
@@ -343,3 +343,20 @@ export function last<T extends any[]>(arr: T): Last<T> {
 export function bnMax<T extends BigNumber>(...args: [T, ...T[]]): T {
   return args.reduce((a, b) => (a.lt(b) ? b : a));
 }
+
+/**
+ * Type helper to recursively map decodable properties to their simpler encoded types;
+ * This allows e.g. types decodable as BigNumbers to be passed in [recursive] properties where
+ * BigNumbers are expected at runtime, as long as the object is decoded/validated before use.
+ */
+export type Decodable<T> = T extends BigNumber
+  ? BigNumberish
+  : T extends string
+  ? string
+  : T extends boolean
+  ? boolean
+  : T extends number
+  ? number
+  : T extends null | symbol
+  ? T
+  : { [K in keyof T]: Decodable<T[K]> };
```

Comment:
Pretty impressive approach. I would love if we could quickly talk about it so I can understand the full scope. Would be good to invite Paul as well. Just a short session.